dsgrid/project.py

```diff
@@ -220,9 +201,151 @@ def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+            logger.info("Use cached project-mapped dataset %s", dataset_id)
+
+        logger.info("Finished processing query for dataset_id=%s", dataset_id)
+        return cached_dataset_path
+
+    def _process_exponential_growth_dataset(
+        self,
+        context: QueryContext,
+        cached_datasets_dir: Path,
+        dataset: ExponentialGrowthDatasetModel,
+    ) -> Path:
+        logger.info("Apply exponential growth for dataset_id=%s", dataset.initial_value_dataset_id)
+        project_version = f"{context.model.project.project_id}__{context.model.project.version}"
+        model_hash, text = context.model.project.dataset.params.serialize()
+        hash_dir = cached_datasets_dir / project_version / model_hash
+        if not hash_dir.exists():
+            hash_dir.mkdir(parents=True)
+            model_file = hash_dir / "model.json"
+            model_file.write_text(text)
+        cached_dataset_path = hash_dir / (dataset.dataset_id + ".parquet")
+        metadata_file = Path(str(cached_dataset_path).replace(".parquet", ".json5"))
+        if try_read_dataframe(cached_dataset_path) is None:
+            self._build_exponential_growth_dataset(
+                context,
+                cached_datasets_dir,
+                dataset,
+                cached_dataset_path,
+                metadata_file,
+            )
+        else:
+            assert metadata_file.exists(), metadata_file
+            context.set_dataset_metadata_from_file(dataset.dataset_id, metadata_file)
+            logger.info("Use cached project-mapped dataset %s", dataset.dataset_id)
+
+        return cached_dataset_path
+
+    @track_timing(timer_stats_collector)
+    def _build_exponential_growth_dataset(
+        self, context, cached_datasets_dir, dataset, dataset_path, metadata_file
+    ):
+        def get_myear_column(dataset_id):
+            names = list(
+                context.get_dimension_query_names(DimensionType.MODEL_YEAR, dataset_id=dataset_id)
+            )
+            assert len(names) == 1, f"{dataset_id=} {names=}"
+            return names[0]
+
+        iv_path = self._process_dataset(
+            context, cached_datasets_dir, dataset.initial_value_dataset_id
+        )
+        gr_path = self._process_dataset(
+            context, cached_datasets_dir, dataset.growth_rate_dataset_id
+        )
+        pivoted_columns = context.get_pivoted_columns(dataset_id=dataset.initial_value_dataset_id)
+        pivoted_columns_gr = context.get_pivoted_columns(dataset_id=dataset.growth_rate_dataset_id)
+        if pivoted_columns != pivoted_columns_gr:
+            raise Exception(
+                f"BUG: Mismatch in pivoted columns: "
+                f"{pivoted_columns.symmetric_difference(pivoted_columns_gr)}"
+            )
+
+        model_year_column = get_myear_column(dataset.initial_value_dataset_id)
+        model_year_column_gr = get_myear_column(dataset.growth_rate_dataset_id)
```

Comment:
How would this only have 1 single model_year?

Reply:
This is the model year column, which is a dimension query name. Or am I missing your question?